### PR TITLE
Flink: Watermark read options

### PR DESCRIPTION
**Pull request by rodmeneses**
1. Adds read option to specify watermark-column for Flink Iceberg connector 
2. Adds read option to specify watermark-column-time-unit for Flink Iceberg connector 
Now it is possible to pass the above parameters in the SQL statement:
```
select * from t /*+ OPTIONS('watermark-column'='t2','watermark-column-time-unit'='MILLISECONDS')*/
```

### CODE REVIEW

File: docs/flink-configuration.md [hunk 1]
```diff
@@ -133,7 +133,8 @@ env.getConfig()
...
+| watermark-timeunit            | connector.iceberg.watermark-timeunit            | N/A                          | TimeUnit.MICROSECONDS            | Specifies the watermark time unit to use for watermark generation.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
```
Comment (pvary) [inline docs/flink-configuration.md:137]: Could we describe to possible values? What should the user write in the option string to get the desired result? Something like: `The possible values are 
DAYS, HOURS, MINUTES, SECONDS, MILLISECONDS, MICROSECONDS, NANOSECONDS`

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/MiniClusterResource.java [hunk 2]
```diff
@@ -26,6 +26,7 @@
 public class MiniClusterResource {
 
   private static final int DEFAULT_TM_NUM = 1;
+
```
Comment (pvary) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/MiniClusterResource.java:29]: nit: not needed, please remove from this change

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java [hunk 3]
```diff
@@ -42,6 +64,7 @@ public TestIcebergSourceBoundedSql(String fileFormat) {
   public void before() throws IOException {
     Configuration tableConf = getTableEnv().getConfig().getConfiguration();
     tableConf.setBoolean(FlinkConfigOptions.TABLE_EXEC_ICEBERG_USE_FLIP27_SOURCE.key(), true);
+    // tableConf.set
```
Comment (pvary) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java:67]: mistake?

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java [hunk 4]
```diff
@@ -77,4 +100,92 @@ protected List<Row> run(
...
+    record.setField("t2", t2);
+    return record;
+  }
+
+  @Test
+  public void testWatermarkOptions() throws Exception {
+    // Skip AVRO since we don't collect metrics for it, and hence we cannot use watermark column as
+    // there are no stats
+    // re: https://github.com/apache/iceberg/pull/1963
+    if (fileFormat != FileFormat.AVRO) {
```
Comment (pvary) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java:116]: Please use `Assume`

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java [hunk 5]
```diff
@@ -77,4 +100,92 @@ protected List<Row> run(
...
+      recordsDataFile2.add(late2);
+      DataFile dataFile2 = helper.writeFile(recordsDataFile2);
+
+      helper.appendToTable(dataFile1, dataFile2);
+
+      // first assertion: early and then late
+      List<Record> expected = Lists.newArrayList();
+      expected.addAll(recordsDataFile1);
+      expected.addAll(recordsDataFile2);
+      TestHelpers.assertRecords(
```
Comment (pvary) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java:152]: Is this test fails, if you change the order of the records?
I expect that `TestHelpers.assertRecords` is independent of the order. See: https://github.com/apache/iceberg/blob/22b95dc70eb46e82a0d51fd4cfb71ad4cbdf946a/flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/TestHelpers.java#L146

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java [hunk 6]
```diff
@@ -39,4 +47,16 @@ public void before() throws IOException {
...
+      new Schema(
+          required(1, "t1", Types.TimestampType.withoutZone()),
+          required(2, "t2", Types.LongType.get()));
+
+  protected Record generateRecord(Instant t1, long t2) {
+    Record record = GenericRecord.create(schema);
+    record.setField("t1", t1.atZone(ZoneId.systemDefault()).toLocalDateTime());
+    record.setField("t2", t2);
+    return record;
+  }
```
Comment (pvary) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java:61]: Is this needed?

File: docs/flink-configuration.md [hunk 7]
```diff
@@ -133,7 +133,8 @@ env.getConfig()
...
+| watermark-timeunit            | connector.iceberg.watermark-timeunit            | N/A                          | TimeUnit.MICROSECONDS            | Specifies the watermark time unit to use for watermark generation. The possible values are  DAYS, HOURS, MINUTES, SECONDS, MILLISECONDS, MICROSECONDS, NANOSECONDS                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
```
Comment (pvary) [inline docs/flink-configuration.md:137]: nit: `.` at the end 😄

File: flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/source/IcebergTableSource.java [hunk 8]
```diff
@@ -131,16 +131,21 @@ private DataStream<RowData> createDataStream(StreamExecutionEnvironment execEnv)
   private DataStreamSource<RowData> createFLIP27Stream(StreamExecutionEnvironment env) {
     SplitAssignerType assignerType =
         readableConfig.get(FlinkConfigOptions.TABLE_EXEC_SPLIT_ASSIGNER_TYPE);
-    IcebergSource<RowData> source =
+    // null check on assignerType
```
Comment (pvary) [inline flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/source/IcebergTableSource.java:134]: I think this comment should be reworded or removed, but after a month nobody will understand what this supposed to mean 😄

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java [hunk 9]
```diff
@@ -77,4 +100,92 @@ protected List<Row> run(
...
   }
+
+  protected Record generateRecord(Instant t1, long t2) {
+    Record record = GenericRecord.create(SCHEMA_TS);
+    record.setField("t1", t1.atZone(ZoneId.systemDefault()).toLocalDateTime());
+    record.setField("t2", t2);
+    return record;
+  }
+
+  @Test
```
Comment (pvary) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java:111]: Please add some comment here that we are testing the order of the splits.

File: flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/source/IcebergTableSource.java (hunk 8 above)
Comment (rodmeneses) [reply flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/source/IcebergTableSource.java:134]: will remove it

File: flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/FlinkConfigOptions.java [hunk 10]
```diff
@@ -94,7 +94,7 @@ private FlinkConfigOptions() {}
   public static final ConfigOption<SplitAssignerType> TABLE_EXEC_SPLIT_ASSIGNER_TYPE =
       ConfigOptions.key("table.exec.iceberg.split-assigner-type")
           .enumType(SplitAssignerType.class)
-          .defaultValue(SplitAssignerType.SIMPLE)
+          .defaultValue(null)
```
Comment (mas-chen) [inline flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/FlinkConfigOptions.java:97]: Is this intended? This would be a breaking change if a user did not provide a split assigner type before

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java [hunk 11]
```diff
@@ -77,4 +100,92 @@ protected List<Row> run(
...
+        run(
+            SCHEMA_TS,
+            ImmutableList.of(),
+            ImmutableMap.of(
+                "watermark-column",
+                "t1",
+                "split-file-open-cost",
+                "128000000",
+                "write-parallelism",
+                "1"),
```
Comment (pvary) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java:163]: Does this also effect the read parallelism too?
Could you please check the effective parallelism in the test?

I would expect we need to use https://nightlies.apache.org/flink/flink-docs-release-1.9/dev/table/config.html#table-exec-resource-default-parallelism to change the parallelism of the reads

File: flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/FlinkConfigOptions.java (hunk 10 above)
Comment (rodmeneses) [reply flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/FlinkConfigOptions.java:97]: good catch. 
I changed that to null, as we had an assertion before:
```
      if (splitAssignerFactory != null) {
        Preconditions.checkArgument(
            watermarkColumn == null,
            "Watermark column and SplitAssigner should not be set in the same source");
      }
```

however, after my work on this PR, it is safe to remove the above assertion, since we will override the factory with a `OrderedSplitAssignerFactory` when the `watermarkColumn` is specified

Comment (mas-chen) [reply flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/FlinkConfigOptions.java:97]: Makes sense. It would be good to update the docs to reveal how the default is determined

Comment (rodmeneses) [reply flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/FlinkConfigOptions.java:97]: I'm updating the doc entry for `watermark-column`:

```
Specifies the watermark column to use for watermark generation. Setting this value will override the `splitAssignerFactory` option.
```
wdyt? @mas-chen @pvary

Comment (mas-chen) [reply flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/FlinkConfigOptions.java]: Sounds good. If users have not opt-ed into the new feature, I would expect their SQL queries to still work. Can you add tests to ensure that?

Also, I think we want to set this to `.noDefaultValue`

Comment (rodmeneses) [reply flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/FlinkConfigOptions.java]: Not according to the code.
```
   if (watermarkColumn != null) {
        // Column statistics is needed for watermark generation
        context = context.copyWithColumnStat(watermarkColumn);
        SplitWatermarkExtractor watermarkExtractor =
            new ColumnStatsWatermarkExtractor(icebergSchema, watermarkColumn, watermarkTimeUnit);
        emitter = SerializableRecordEmitter.emitterWithWatermark(watermarkExtractor);
        splitAssignerFactory =
            new OrderedSplitAssignerFactory(SplitComparators.watermark(watermarkExtractor));
      }
```

Comment (rodmeneses) [reply flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/FlinkConfigOptions.java]: cc: @pvary

Comment (pvary) [reply flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/FlinkConfigOptions.java]: If the splits are not ordered, then we will have fluctuating watermarks. We do not emit those, which are not in order, but beats the purpose of the whole watermark generation feature.
Imagine a situation where we reading time series data, and read the latest file first. Every other file will contain late data in this case, and might be dropped.

So while technically possible, I rather not allow the users to shoot themselves in the foot.

Comment (rodmeneses) [reply flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/FlinkConfigOptions.java]: > If users have not opt-ed into the new feature, I would expect their SQL queries to still work. Can you add tests to ensure that?

Hi! I didn't get your suggestion about the unit test. Would you please rephrase? Thanks @mas-chen

File: flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/source/IcebergTableSource.java [hunk 12]
```diff
@@ -131,16 +131,20 @@ private DataStream<RowData> createDataStream(StreamExecutionEnvironment execEnv)
...
-            .assignerFactory(assignerType.factory())
             .properties(properties)
             .project(getProjectedSchema())
             .limit(limit)
             .filters(filters)
-            .flinkConfig(readableConfig)
-            .build();
+            .flinkConfig(readableConfig);
+
+    if (assignerType != null) {
```
Comment (stevenzwu) [inline flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/source/IcebergTableSource.java:143]: why is this changed? assigner type has a default type. old code should be fine, right?

File: flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/source/ScanContext.java [hunk 13]
```diff
@@ -298,6 +313,40 @@ public ScanContext copyWithAppendsBetween(Long newStartSnapshotId, long newEndSn
         .planParallelism(planParallelism)
         .maxPlanningSnapshotCount(maxPlanningSnapshotCount)
         .maxAllowedPlanningFailures(maxAllowedPlanningFailures)
+        .watermarkColumn(watermarkColumn)
+        .watermarkTimeUnit(watermarkTimeUnit)
+        .build();
+  }
+
+  public ScanContext copyWithColumnStat(String columnStat) {
```
Comment (stevenzwu) [inline flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/source/ScanContext.java:321]: related to earlier sure. this may not be needed

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/TestHelpers.java [hunk 14]
```diff
@@ -146,6 +159,10 @@ public static void assertRows(List<Row> results, List<Row> expected) {
     Assertions.assertThat(results).containsExactlyInAnyOrderElementsOf(expected);
   }
 
+  public static void assertRowsWithOrder(List<Row> results, List<Row> expected) {
+    Assertions.assertThat(results).isEqualTo(expected);
```
Comment (stevenzwu) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/TestHelpers.java:163]: use `containsExactly`?

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/TestHelpers.java [hunk 15]
```diff
@@ -126,6 +126,19 @@ public static List<Row> convertRowDataToRow(List<RowData> rowDataList, RowType r
         .collect(Collectors.toList());
   }
 
+  public static void assertRecordsWithOrder(
+      List<Row> results, List<Record> expectedRecords, Schema schema) {
+    List<Row> expected = Lists.newArrayList();
+    @SuppressWarnings("unchecked")
```
Comment (stevenzwu) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/TestHelpers.java:132]: we can extract a method out to avoid duplication with the method below

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java [hunk 16]
```diff
@@ -77,4 +102,88 @@ protected List<Row> run(
...
+  /**
+   * Tests the order of splits returned when setting the watermark-column and watermark-timeunit
+   * options
+   */
+  @Test
+  public void testWatermarkOptions() throws Exception {
+    // Skip AVRO since we don't collect metrics for it, and hence we cannot use watermark column as
+    // there are no stats
+    // re: https://github.com/apache/iceberg/pull/1963
+    Assume.assumeTrue("Temporary skip AVRO", FileFormat.AVRO != fileFormat);
```
Comment (stevenzwu) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java:122]: this new test is probably better added to TestIcebergSourceSql.

the base class for TestIcebergSourceBoundedSql also test different file format, which is not necessary for the advanced options like watermark alignment

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java [hunk 17]
```diff
@@ -77,4 +102,88 @@ protected List<Row> run(
...
+
+    Table table = catalogResource.catalog().createTable(TestFixtures.TABLE_IDENTIFIER, SCHEMA_TS);
+    long baseTime = 1702382109000L;
```
Comment (stevenzwu) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java:125]: curious why is baseTime hardcoded? why not `System.currentTimeMillis`?

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java [hunk 18]
```diff
@@ -77,4 +102,88 @@ protected List<Row> run(
...
+
+    GenericAppenderHelper helper = new GenericAppenderHelper(table, fileFormat, TEMPORARY_FOLDER);
+
+    // File 1 - early timestamps, new longs
+    Record early1 =
+        generateRecord(Instant.ofEpochMilli(baseTime), baseTime + (1000 * 60 * 60 * 24 * 30L));
```
Comment (stevenzwu) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java:131]: why do use different values for t1 and t2? wasn't it more intuitive if they are the same?

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java [hunk 19]
```diff
@@ -77,4 +102,88 @@ protected List<Row> run(
...
+            "",
+            "*"),
+        expected,
+        SCHEMA_TS);
+
+    // second assertion: late and then early
+    expected.clear();
+    expected.addAll(recordsDataFile2);
+    expected.addAll(recordsDataFile1);
+    TestHelpers.assertRecordsWithOrder(
```
Comment (stevenzwu) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java:173]: this should be separated out to another test method.

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java [hunk 20]
```diff
@@ -39,4 +62,76 @@ public void before() throws IOException {
...
+
+  @Test
+  public void testWatermarkOptionsDescending() throws Exception {
+    List<Record> expected = testWatermarkOptionsInternal(false);
+    TestHelpers.assertRecordsWithOrder(
+        run(
+            ImmutableMap.of(
+                "watermark-column",
+                "t2",
+                "watermark-time-unit",
```
Comment (pvary) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java:128]: Shouldn't this be `watermark-column-time-unit`?

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java [hunk 21]
```diff
@@ -39,4 +62,76 @@ public void before() throws IOException {
         .getConfiguration()
         .set(TableConfigOptions.TABLE_DYNAMIC_TABLE_OPTIONS_ENABLED, true);
   }
+
+  private Record generateRecord(Instant t1) {
+    Record record = GenericRecord.create(SCHEMA_TS);
+    record.setField("t1", t1.atZone(ZoneId.systemDefault()).toLocalDateTime());
+    record.setField("t2", t1.getEpochSecond());
```
Comment (pvary) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java:69]: What I do not like in this test, that if we make a mistake and switch `t1` with `t2` in somewhere in the production code, or in the test then we will not see a difference in the test results.

That is why I originally suggested, that the different columns have different values to make sure that the `watermark-column` value is propagated correctly

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java [hunk 22]
```diff
@@ -39,4 +62,76 @@ public void before() throws IOException {
...
+            ImmutableMap.of("watermark-column", "t1", "split-file-open-cost", "128000000"),
+            "",
+            "*"),
+        expected,
+        SCHEMA_TS);
+  }
+
+  @Test
+  public void testWatermarkOptionsDescending() throws Exception {
+    List<Record> expected = testWatermarkOptionsInternal(false);
```
Comment (stevenzwu) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java:122]: I don't quite understand here. both test methods enable watermark alignment. one get the ascending order. another get the reverse order.

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java (hunk 21 above)
Comment (pvary) [reply flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java:69]: To make sure that the watermark generation is set, the test checks that the ordering of the splits is correct. We create 2 files add them to the table. If the watermark generation is not set, then the records from the files are returned in the same order as we added to the table. If the watermark generation is set, then the splits are returned in the order based on the data.

If the data is the same in `t1` and in `t2` columns, then there is no difference when we order by (generate watermarks) based on `t1` or `t2`.  So the test is less restrictive.

Comment (stevenzwu) [reply flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java:69]: > If the watermark generation is not set, then the records from the files are returned in the same order as we added to the table. If the watermark generation is set, then the splits are returned in the order based on the data.

both test methods enable watermark generation. 

> If the data is the same in t1 and in t2 columns, then there is no difference when we order by (generate watermarks) based on t1 or t2. So the test is less restrictive.

I guess I probably understand why previously t1 and t2 are set to different values. you want the order to be different depnding on t1 or t2 is the watermark column.

Comment (rodmeneses) [reply flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java]: the unit test is using the 2 columns approach, so that we can cover the scenarios described by @pvary

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java (hunk 22 above)
Comment (rodmeneses) [reply flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java]: the `generateExpectedRecords` (which is now renamed, before it used to be `testWatermarkOptionsInternal`) method will generate records in either ascending or descending order, with respect to their data file. For each direction, we have a different unit test

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java [hunk 23]
```diff
@@ -18,25 +18,139 @@
...
     SqlHelpers.sql(
-        getTableEnv(),
+        tableEnvironment,
         "create catalog iceberg_catalog with ('type'='iceberg', 'catalog-type'='hadoop', 'warehouse'='%s')",
         catalogResource.warehouse());
-    SqlHelpers.sql(getTableEnv(), "use catalog iceberg_catalog");
-    getTableEnv()
+    SqlHelpers.sql(tableEnvironment, "use catalog iceberg_catalog");
+
+    tableEnvironment
```
Comment (stevenzwu) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java:65]: this can be changed to using `tableConf`

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java [hunk 24]
```diff
@@ -18,25 +18,139 @@
...
+  /** Generates the records in the expected order, with respect to their datafile */
+  private List<Record> generateExpectedRecords(boolean ascending) throws Exception {
+    Table table = catalogResource.catalog().createTable(TestFixtures.TABLE_IDENTIFIER, SCHEMA_TS);
+    long baseTime = 1702382109000L;
+
+    GenericAppenderHelper helper =
+        new GenericAppenderHelper(table, FileFormat.PARQUET, TEMPORARY_FOLDER);
+
+    // File 1 - early timestamps, new longs
+    Record early1 =
```
Comment (stevenzwu) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java:87]: the naming of `early` vs `late` is inaccurate, as the two columns have different time order.

File: flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/FlinkReadConf.java [hunk 25]
```diff
@@ -190,4 +191,22 @@ public int maxAllowedPlanningFailures() {
...
+  public String watermarkColumn() {
+    return confParser
+        .stringConf()
+        .option(FlinkReadOptions.WATERMARK_COLUMN)
+        .flinkConfig(FlinkReadOptions.WATERMARK_COLUMN_OPTION)
+        .defaultValue(FlinkReadOptions.WATERMARK_COLUMN_OPTION.defaultValue())
+        .parseOptional();
+  }
+
+  public TimeUnit watermarkTimeUnit() {
```
Comment (stevenzwu) [inline flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/FlinkReadConf.java:204]: method name should be `watermarkColumnTimeUnit`

File: flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/source/ScanContext.java [hunk 26]
```diff
@@ -272,6 +279,14 @@ public int maxAllowedPlanningFailures() {
     return maxAllowedPlanningFailures;
   }
 
+  public String watermarkColumn() {
+    return watermarkColumn;
+  }
+
+  public TimeUnit watermarkTimeUnit() {
```
Comment (stevenzwu) [inline flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/source/ScanContext.java:286]: method name should be `watermarkColumnTimeUnit`

**Review by stevenzwu — APPROVED** [review]

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java [hunk 27]
```diff
@@ -18,25 +18,136 @@
...
+        generateRecord(Instant.ofEpochMilli(baseTime), baseTime + (1000 * 60 * 60 * 24 * 30L));
+    Record file1Record2 =
+        generateRecord(
+            Instant.ofEpochMilli(baseTime - 10 * 1000L), baseTime + (1000 * 60 * 60 * 24 * 35L));
+
+    List<Record> recordsDataFile1 = Lists.newArrayList();
+    recordsDataFile1.add(file1Record1);
+    recordsDataFile1.add(file1Record2);
+    DataFile dataFile1 = helper.writeFile(recordsDataFile1);
+    // File 2 - old timestamps, old longs
```
Comment (pvary) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java:94]: This comment is not correct:
```
// File 2 - late timestamps, old longs
```

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java [hunk 28]
```diff
@@ -18,25 +18,136 @@
...
+            Instant.ofEpochMilli(baseTime + 14 * 1000L), baseTime - (1000 * 60 * 60 * 24 * 30L));
+    Record file2Record2 =
+        generateRecord(
+            Instant.ofEpochMilli(baseTime + 12 * 1000L), baseTime - (1000 * 60 * 61 * 24 * 35L));
+
+    List<Record> recordsDataFile2 = Lists.newArrayList();
+    recordsDataFile2.add(file2Record1);
+    recordsDataFile2.add(file2Record2);
+
+    // early1 - early2 -- late1 late 2
```
Comment (pvary) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceSql.java:106]: I do not get this comment.
Maybe something like this - feel free to reword if you feel so:
```
// Expected records if the splits are ordered
//     - ascending (watermark from t1) - records from the split with early timestamps then records from the split with late timestamps
//     - descending (watermark from t2) - records from the split with old longs then records from the split with new longs
```

**Review by pvary — APPROVED** [review]
Left some minor comments for comments 😄 
Otherwise +1 LGTM

**Comment by pvary** [comment]
Thanks @rodmeneses for the PR, and @stevenzwu and @mas-chen for the review!
